sc2/bot_ai_internal.py

```diff
@@ -177,14 +177,154 @@ def expansion_locations(self) -> dict[Point2, Units]:
         )
         return self.expansion_locations_dict
 
+    def _cluster_center(self, group: list[Unit]) -> Point2:
+        """
+        Calculates the geometric center (centroid) of a given group of units.
+
+        Parameters:
+        group: A list of Unit objects representing the group of units for
+                            which the center is to be calculated.
+
+        Raises:
+        ValueError: If the provided group is empty.
+
+        Returns:
+        Point2: The calculated centroid of the group as a Point2 object.
+        """
+        if not group:
+            raise ValueError("Cannot calculate center of empty group")
+
+        total_x = total_y = 0
+        for unit in group:
+            total_x += unit.position.x
+            total_y += unit.position.y
+
+        count = len(group)
+        return Point2((total_x / count, total_y / count))
+
+    def _find_expansion_location(
+        self, resources: Units | list[Unit], amount: int, offsets: list[tuple]
+    ) -> Point2 | None:
+        """
+        Finds the most suitable expansion location for resources.
+
+        Parameters:
+            resources: The list of resource entities or units near which the
+                                             expansion location needs to be found.
+            amount: The total number of resource entities or units to consider.
+            offsets (list[tuple]): A list of coordinate pairs denoting position offsets to consider
+                                   around the center of resources.
```

Comment:
```suggestion
    def _find_expansion_location(
        self, resources: Units | list[Unit], amount: int, offsets: list[tuple[float, float]]
    ) -> Point2:
        """
        Finds the most suitable expansion location for resources.

        Parameters:
            resources: The list of resource entities or units near which the
                                             expansion location needs to be found.
            amount: The total number of resource entities or units to consider.
            offsets (list[tuple[float, float]]): A list of coordinate pairs denoting position offsets to consider
                                   around the center of resources.
```
I guess the tuples can be `Point2` too and it should still work.
Also I don't see a `return None` so that the return type can be reduced to `Point2`? In the othe code calling this function, you never expect it to return `None`.

Reply:
Good spot! Not sure why it had a `None` return type